File: problems/basic/819/solution.ts
// Accepted: 66ms (58.46%), 45.90MB (30.77%)

function mostCommonWord(paragraph: string, banned: string[]): string {
    const bannedSet = new Set(banned);
    const matches = Array.from(paragraph.toLowerCase().matchAll(/[a-z]+/g), ([match]) => match);

    const counts = new Map<string, number>();
    for (const match of matches) {
        counts.set(match, (counts.get(match) ?? 0) + 1);
    }

    let mostCommonWord: string;
    let mostCommonCount = 0;
    for (const [word, count] of counts.entries()) {
        if (bannedSet.has(word)) {
            continue;
        }
        if (count <= mostCommonCount) {
            continue;
        }

        mostCommonWord = word;
        mostCommonCount = count;
    }

    return mostCommonWord!;
};

File: problems/basic/819/solution.cpp
// Accepted: 3ms (81.58%), 9.77MB (5.42%)

class Solution {
public:
    string mostCommonWord(string paragraph, vector<string>& banned) {
        transform(paragraph.begin(), paragraph.end(), paragraph.begin(), [](const unsigned char& ch) {
            return isalpha(ch) ? (unsigned char)tolower(ch) : ' ';
        });

        vector<string> words;
        stringstream ss(paragraph);
        string s;
        while (ss >> s) {
            words.push_back(s);
        }

        unordered_map<string, int> counts;
        for (string word : words) {
            counts[word]++;
        }

        unordered_set<string> bannedSet(banned.begin(), banned.end());

        string mostCommonWord;
        int mostCommonCount = 0;
        for (auto& [word, count] : counts) {
            if (bannedSet.count(word) != 0) {
                continue;
            }
            if (count <= mostCommonCount) {
                continue;
            }

            mostCommonCount = count;
            mostCommonWord = word;
        }

        return mostCommonWord;
    }
};
